File: E-paper_Separate_Program/10.85inch_e-Paper_G/RaspberryPi/python/lib/epd10in85g.py
# /*****************************************************************************
# * | File        :	  epd12in48.py
# * | Author      :   Waveshare electrices
# * | Function    :   Hardware underlying interface
# * | Info        :
# *----------------
# * |	This version:   V1.0
# * | Date        :   2019-11-01
# * | Info        :   
# ******************************************************************************/
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documnetation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to  whom the Software is
# furished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS OR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
#
import time
import epdconfig

import PIL
from PIL import Image
import io

EPD_WIDTH       = 1360//2
EPD_HEIGHT      = 480

class EPD():
    def __init__(self):
        self.width = EPD_WIDTH
        self.height = EPD_HEIGHT
        self.BLACK  = 0x000000   #   00  BGR
        self.WHITE  = 0xffffff   #   01
        self.YELLOW = 0x00ffff   #   10
        self.RED    = 0x0000ff   #   11
        
        self.EPD_CS_M_PIN  = epdconfig.EPD_CS_M_PIN
        self.EPD_CS_S_PIN  = epdconfig.EPD_CS_S_PIN

        self.EPD_DC_PIN  = epdconfig.EPD_DC_PIN
        self.EPD_RST_PIN  = epdconfig.EPD_RST_PIN
        self.EPD_BUSY_PIN  = epdconfig.EPD_BUSY_PIN
        self.EPD_PWR_PIN  = epdconfig.EPD_PWR_PIN


    
    def Reset(self):
        epdconfig.digital_write(self.EPD_RST_PIN, 1)
        epdconfig.delay_ms(200) 
        epdconfig.digital_write(self.EPD_RST_PIN, 0)         # module reset
        epdconfig.delay_ms(2)
        epdconfig.digital_write(self.EPD_RST_PIN, 1)
        epdconfig.delay_ms(200)   

    def CS_ALL(self, Value):
        epdconfig.digital_write(self.EPD_CS_M_PIN, Value)
        epdconfig.digital_write(self.EPD_CS_S_PIN, Value)

    def SendCommand(self, Command):
        epdconfig.digital_write(self.EPD_DC_PIN, 0)
        epdconfig.spi_writebyte(Command)

    def SendData(self, Data):
        epdconfig.digital_write(self.EPD_DC_PIN, 1)
        epdconfig.spi_writebyte(Data)
    
    def SendData2(self, buf, Len):
        epdconfig.digital_write(self.EPD_DC_PIN, 1)
        epdconfig.spi_writebyte2(buf, Len)

    def ReadBusyH(self):
        print("e-Paper busy H")
        epdconfig.delay_ms(100)
        while(epdconfig.digital_read(self.EPD_BUSY_PIN) == 0):      # 0: idle, 1: busy
            epdconfig.delay_ms(20)
        print("e-Paper busy release")

    def TurnOnDisplay(self):
        self.CS_ALL(0)
        self.SendCommand(0x12)
        self.SendData(0x00)
        self.CS_ALL(1)
        self.ReadBusyH()

    def Init(self):
        print("EPD init...")
        epdconfig.module_init()
        
        
        self.Reset() 
        self.ReadBusyH()

        self.CS_ALL(0)

        self.SendCommand(0x4D)
        self.SendData(0x78)	
       

        self.SendCommand(0x00)
        self.SendData(0x2F)	
        self.SendData(0x29)	

        self.SendCommand(0x06)
        self.SendData(0x0d)
        self.SendData(0x12)
        self.SendData(0x30)
        self.SendData(0x20)
        self.SendData(0x19)
        self.SendData(0x3D)
        self.SendData(0x0C)

        self.SendCommand(0x06)
        self.SendData(0x0d)
        self.SendData(0x12)
        self.SendData(0x30)
        self.SendData(0x20)
        self.SendData(0x19)
        self.SendData(0x3D)
        self.SendData(0x0C)

        self.SendCommand(0x50)
        self.SendData(0x37)	

        self.SendCommand(0x61)
        self.SendData(int(self.width/256))
        self.SendData(self.width%256)
        self.SendData(int(self.height/256))
        self.SendData(self.height%256)

        self.SendCommand(0x65)
        self.SendData(0x00)	
        self.SendData(0x00)	
        self.SendData(0x00)	
        self.SendData(0x00)	

        self.SendCommand(0xE0)
        self.SendData(0x01)

        self.SendCommand(0xE3)
        self.SendData(0x08)  

        self.SendCommand(0xE5)
        self.SendData(0x08) 

        self.SendCommand(0xE9)
        self.SendData(0x01)   

        self.SendCommand(0x04)
        self.ReadBusyH()
        self.CS_ALL(1)
    
    def getbuffer(self, image):
        # Create a pallette with the 4 colors supported by the panel
        pal_image = Image.new("P", (1,1))
        pal_image.putpalette( (0,0,0,  255,255,255,  255,255,0,   255,0,0) + (0,0,0)*252)

        # Check if we need to rotate the image
        imwidth, imheight = image.size
        if(imwidth == self.width*2 and imheight == self.height):
            image_temp = image
        elif(imwidth == self.height and imheight == self.width*2):
            image_temp = image.rotate(90, expand=True)
        else:
            logger.warning("Invalid image dimensions: %d x %d, expected %d x %d" % (imwidth, imheight, self.width*2, self.height))

        # Convert the soruce image to the 4 colors, dithering if needed
        image_4color = image_temp.convert("RGB").quantize(palette=pal_image)
        buf_4color = bytearray(image_4color.tobytes('raw'))

        # into a single byte to transfer to the panel
        if (self.width*2) % 4 == 0 :
            Width = (self.width*2) // 4
        else :
            Width = (self.width*2) // 4 + 1
        Height = self.height 
        buf = [0x00] * int(Width * Height)
        idx = 0
        for j in range(0, Height):
            for i in range(0, Width):
                    buf[i + j * Width] = (buf_4color[idx] << 6) + (buf_4color[idx+1] << 4) + (buf_4color[idx+2] << 2) + buf_4color[idx+3]
                    idx = idx + 4
        return buf
    
    def Clear(self, color=0x55):
        epdconfig.digital_write(self.EPD_CS_M_PIN, 0)
        self.SendCommand(0x10)
        for i in range(self.height):
            self.SendData2([color]* int(self.width/2), int(self.width/2))
        self.CS_ALL(1)
        epdconfig.digital_write(self.EPD_CS_S_PIN, 0)
        self.SendCommand(0x10)
        for i in range(self.height):
            self.SendData2([color]* int(self.width/2), int(self.width/2))
        self.CS_ALL(1)

        self.TurnOnDisplay()

    def display(self, image):
        Width =int(self.width / 4)
        Width1 =int(self.width / 2)

        epdconfig.digital_write(self.EPD_CS_M_PIN, 0)
        self.SendCommand(0x10)
        for i in range(self.height):
            self.SendData2(image[i * Width1 : i * Width1+Width], Width)
        self.CS_ALL(1)

        epdconfig.digital_write(self.EPD_CS_S_PIN, 0)
        self.SendCommand(0x10)
        for i in range(self.height):
            self.SendData2(image[i * Width1+Width : i * Width1+Width1], Width)
        self.CS_ALL(1)

        self.TurnOnDisplay()

    def sleep(self):
        self.CS_ALL(0)
        self.SendCommand(0x02) # POWER_OFF
        self.SendData(0X00)
        epdconfig.delay_ms(100)
        
        self.SendCommand(0x07) # DEEP_SLEEP
        self.SendData(0XA5)
        self.CS_ALL(1)

        epdconfig.delay_ms(2000)
        epdconfig.module_exit()
### END OF FILE ###

File: E-paper_Separate_Program/10.85inch_e-Paper_G/ESP32/EPD_10in85g.cpp
/*****************************************************************************
* | File      	:  	EPD_10in85g.c
* | Author      :   Waveshare team
* | Function    :   10.85inch e-paper (G)
* | Info        :
*----------------
* |	This version:   V1.0
* | Date        :   2024-08-19
* | Info        :
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documnetation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to  whom the Software is
# furished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS OR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
#
******************************************************************************/
#include "EPD_10in85g.h"
#include "Debug.h"

/******************************************************************************
function :	Software reset
parameter:
******************************************************************************/
static void EPD_10in85g_Reset(void)
{
    DEV_Digital_Write(EPD_RST_PIN, 1);
    DEV_Delay_ms(20);
    DEV_Digital_Write(EPD_RST_PIN, 0);
    DEV_Delay_ms(10);
    DEV_Digital_Write(EPD_RST_PIN, 1);
    DEV_Delay_ms(20);
}

/******************************************************************************
function :	send command
parameter:
     Reg : Command register
******************************************************************************/
static void EPD_10in85g_SendCommand_0(UBYTE Reg)
{
    DEV_Digital_Write(EPD_DC_PIN, 0);
    DEV_Digital_Write(EPD_CS_M_PIN, 0);
    DEV_SPI_WriteByte(Reg);
    DEV_Digital_Write(EPD_CS_M_PIN, 1);
}
static void EPD_10in85g_SendCommand_1(UBYTE Reg)
{
    DEV_Digital_Write(EPD_DC_PIN, 0);
    DEV_Digital_Write(EPD_CS_S_PIN, 0);
    DEV_SPI_WriteByte(Reg);
    DEV_Digital_Write(EPD_CS_S_PIN, 1);
}
static void EPD_10in85g_SendCommand_ALL(UBYTE Reg)
{
    DEV_Digital_Write(EPD_DC_PIN, 0);
    DEV_Digital_Write(EPD_CS_M_PIN, 0);
    DEV_Digital_Write(EPD_CS_S_PIN, 0);
    DEV_SPI_WriteByte(Reg);
    DEV_Digital_Write(EPD_CS_M_PIN, 1);
    DEV_Digital_Write(EPD_CS_S_PIN, 1);
}

/******************************************************************************
function :	send data
parameter:
    Data : Write data
******************************************************************************/

static void EPD_10in85g_SendData_0(UBYTE Data)
{
    DEV_Digital_Write(EPD_DC_PIN, 1);
    DEV_Digital_Write(EPD_CS_M_PIN, 0);
    DEV_SPI_WriteByte(Data);
    DEV_Digital_Write(EPD_CS_M_PIN, 1);
}
static void EPD_10in85g_SendData_1(UBYTE Data)
{
    DEV_Digital_Write(EPD_DC_PIN, 1);
    DEV_Digital_Write(EPD_CS_S_PIN, 0);
    DEV_SPI_WriteByte(Data);
    DEV_Digital_Write(EPD_CS_S_PIN, 1);
}

static void EPD_10in85g_SendnData_0(UBYTE *Data, UDOUBLE Len)
{
    DEV_Digital_Write(EPD_DC_PIN, 1);
    DEV_Digital_Write(EPD_CS_M_PIN, 0);
    DEV_SPI_Write_nByte(Data, Len);
    DEV_Digital_Write(EPD_CS_M_PIN, 1);
}
static void EPD_10in85g_SendnData_1(UBYTE *Data, UDOUBLE Len)
{
    DEV_Digital_Write(EPD_DC_PIN, 1);
    DEV_Digital_Write(EPD_CS_S_PIN, 0);
    DEV_SPI_Write_nByte(Data, Len);
    DEV_Digital_Write(EPD_CS_S_PIN, 1);
}

static void EPD_10in85g_SendData_ALL(UBYTE Data)
{
    DEV_Digital_Write(EPD_DC_PIN, 1);
    DEV_Digital_Write(EPD_CS_M_PIN, 0);
    DEV_Digital_Write(EPD_CS_S_PIN, 0);
    DEV_SPI_WriteByte(Data);
    DEV_Digital_Write(EPD_CS_M_PIN, 1);
    DEV_Digital_Write(EPD_CS_S_PIN, 1);
}

/******************************************************************************
function :	Wait until the busy_pin goes LOW
parameter:
******************************************************************************/
void EPD_10in85g_ReadBusy(void)
{
    Debug("e-Paper busy\r\n");
    // DEV_Delay_ms(1000);
	while(1)
	{	 //=1 BUSY
		if(DEV_Digital_Read(EPD_BUSY_PIN)==1) 
			break;
		DEV_Delay_ms(10);
	}
	DEV_Delay_ms(10);
    Debug("e-Paper busy release\r\n");
}

/******************************************************************************
function :	Turn On Display
parameter:
******************************************************************************/
static void EPD_10in85g_TurnOnDisplay(void)
{

	EPD_10in85g_SendCommand_ALL(0x12);
    EPD_10in85g_SendData_ALL(0x00);	
	EPD_10in85g_ReadBusy();
}

/******************************************************************************
function :	Initialize the e-Paper register
parameter:
******************************************************************************/
void EPD_10in85g_Init(void)
{
	EPD_10in85g_Reset();
    EPD_10in85g_ReadBusy();

    EPD_10in85g_SendCommand_ALL(0x4D);
    EPD_10in85g_SendData_ALL(0x78);	

    EPD_10in85g_SendCommand_ALL(0x00);	//0x00
    EPD_10in85g_SendData_ALL(0x2F);	
    EPD_10in85g_SendData_ALL(0x29);	

    EPD_10in85g_SendCommand_ALL(0x06);//47uH
    EPD_10in85g_SendData_ALL(0x0d);
    EPD_10in85g_SendData_ALL(0x12);
    EPD_10in85g_SendData_ALL(0x30);
    EPD_10in85g_SendData_ALL(0x20);
    EPD_10in85g_SendData_ALL(0x19);
    EPD_10in85g_SendData_ALL(0x3D);
    EPD_10in85g_SendData_ALL(0x0C);

    EPD_10in85g_SendCommand_ALL(0x06);	//0x06
    EPD_10in85g_SendData_ALL(0x0d);
    EPD_10in85g_SendData_ALL(0x12);
    EPD_10in85g_SendData_ALL(0x30);
    EPD_10in85g_SendData_ALL(0x20);
    EPD_10in85g_SendData_ALL(0x19);
    EPD_10in85g_SendData_ALL(0x3D);
    EPD_10in85g_SendData_ALL(0x0C);

    EPD_10in85g_SendCommand_ALL(0x50);	//0x50
    EPD_10in85g_SendData_ALL(0x37);	

    EPD_10in85g_SendCommand_ALL(0x61);//0x61	
    EPD_10in85g_SendData_ALL(EPD_10in85g_WIDTH/256);	
    EPD_10in85g_SendData_ALL(EPD_10in85g_WIDTH%256);	
    EPD_10in85g_SendData_ALL(EPD_10in85g_HEIGHT/256);	
    EPD_10in85g_SendData_ALL(EPD_10in85g_HEIGHT%256);	

    EPD_10in85g_SendCommand_ALL(0x65);	//0x65
    EPD_10in85g_SendData_ALL(0x00);	
    EPD_10in85g_SendData_ALL(0x00);	
    EPD_10in85g_SendData_ALL(0x00);	
    EPD_10in85g_SendData_ALL(0x00);	

    EPD_10in85g_SendCommand_ALL(0xE0);
    EPD_10in85g_SendData_ALL(0x01);

    EPD_10in85g_SendCommand_ALL(0xE3);
    EPD_10in85g_SendData_ALL(0x08);  

    EPD_10in85g_SendCommand_ALL(0xE5);
    EPD_10in85g_SendData_ALL(0x08); 

    EPD_10in85g_SendCommand_ALL(0xE9);
    EPD_10in85g_SendData_ALL(0x01);   

    EPD_10in85g_SendCommand_ALL(0x04); //Power on
    EPD_10in85g_ReadBusy();          //waiting for the electronic paper IC to release the idle signal
}

/******************************************************************************
function :	Clear screen
parameter:
******************************************************************************/
void EPD_10in85g_Clear(UBYTE color)
{
	UWORD Width, Height;
    Width = (EPD_10in85g_WIDTH % 4 == 0)? (EPD_10in85g_WIDTH / 4 ): (EPD_10in85g_WIDTH / 4 + 1);
    Height = EPD_10in85g_HEIGHT;
    UBYTE Color = (color << 6) | (color << 4) | (color << 2) | color;

    EPD_10in85g_SendCommand_0(0x10);
    for (UWORD j = 0; j < Height; j++) {
        for (UWORD i = 0; i < Width; i++) {
            EPD_10in85g_SendData_0(Color);
        }
    }	

    EPD_10in85g_SendCommand_1(0x10);
    for (UWORD j = 0; j < Height; j++) {
        for (UWORD i = 0; i < Width; i++) {
            EPD_10in85g_SendData_1(Color);
        }
    }	

	EPD_10in85g_TurnOnDisplay();
}

/******************************************************************************
function :	Sends the image buffer in RAM to e-Paper and displays
parameter:
	Image : Image data
******************************************************************************/
void EPD_10in85g_Display(const UBYTE *Image)
{
	UWORD Width, Height;
    Width = (EPD_10in85g_WIDTH % 4 == 0)? (EPD_10in85g_WIDTH / 4 ): (EPD_10in85g_WIDTH / 4 + 1);
    Height = EPD_10in85g_HEIGHT;
	
    EPD_10in85g_SendCommand_0(0x10);
    for (UWORD j = 0; j < Height; j++) {
        for (UWORD i = 0; i < Width; i++) {
            EPD_10in85g_SendData_0(Image[j*2*Width + i]);
        }
    }	

    EPD_10in85g_SendCommand_1(0x10);
    for (UWORD j = 0; j < Height; j++) {
        for (UWORD i = 0; i < Width; i++) {
            EPD_10in85g_SendData_1(Image[j*2*Width + i + Width]);
        }
    }

	EPD_10in85g_TurnOnDisplay();	
}

void EPD_10in85g_Display_2(const UBYTE *Image)
{
	UWORD Width, Height;
    Width = (EPD_10in85g_WIDTH % 4 == 0)? (EPD_10in85g_WIDTH / 4 ): (EPD_10in85g_WIDTH / 4 + 1);
    Height = EPD_10in85g_HEIGHT/4;
	
    EPD_10in85g_SendCommand_0(0x10);
    for (UWORD j = 0; j < Height; j++) {
        for (UWORD i = 0; i < Width; i++) {
            EPD_10in85g_SendData_0(Image[j*Width + i]);
        }
    }	
    for (UWORD j = 0; j < Height*3; j++) {
        for (UWORD i = 0; i < Width; i++) {
            EPD_10in85g_SendData_0(0x55);
        }
    }

    EPD_10in85g_SendCommand_1(0x10);
    for (UWORD j = 0; j < Height; j++) {
        for (UWORD i = 0; i < Width; i++) {
            EPD_10in85g_SendData_1(Image[j*Width + i]);
        }
    }
    for (UWORD j = 0; j < Height*3; j++) {
        for (UWORD i = 0; i < Width; i++) {
            EPD_10in85g_SendData_1(0x55);
        }
    }

	EPD_10in85g_TurnOnDisplay();	
}

void EPD_10in85g_DisplayPart(const UBYTE *Image, UWORD xstart, UWORD ystart, UWORD image_width, UWORD image_heigh)
{
    UDOUBLE Width, Height;
    Width = (EPD_10in85g_WIDTH % 4 == 0)? (EPD_10in85g_WIDTH / 4 ): (EPD_10in85g_WIDTH / 4 + 1);
    Height = EPD_10in85g_HEIGHT;
    
    UWORD Xend = ((xstart + image_width)%4 == 0)?((xstart + image_width) / 4 - 1): ((xstart + image_width) / 4 );
    UWORD Yend = ystart + image_heigh-1;
    xstart = xstart / 4;
    
    if(xstart > 170 )
    {
        Xend = Xend - 170;
        xstart = xstart - 170;
        EPD_10in85g_SendCommand_0(0x10);
        for (UDOUBLE i = 0; i < Height; i++) {
            for (UDOUBLE j = 0; j < Width; j++) {
                EPD_10in85g_SendData_0(0x55);
            }
        }

        EPD_10in85g_SendCommand_1(0x10);
        for (UDOUBLE i = 0; i < Height; i++) {
            for (UDOUBLE j = 0; j < Width; j++) {
                if((i<Yend) && (i>=ystart) && (j<Xend) && (j>=xstart)) {
                    EPD_10in85g_SendData_1(Image[(j-xstart) + (image_width/4*(i-ystart))]);
                }
                else
                    EPD_10in85g_SendData_1(0x55);
            }
        }

    }
    else if(Xend < 170 )
    {
        EPD_10in85g_SendCommand_0(0x10);
        for (UDOUBLE i = 0; i < Height; i++) {
            for (UDOUBLE j = 0; j < Width; j++) {
                if((i<Yend) && (i>=ystart) && (j<Xend) && (j>=xstart)) {
                    EPD_10in85g_SendData_0(Image[(j-xstart) + (image_width/4*(i-ystart))]);
                }
                else
                    EPD_10in85g_SendData_0(0x55);
            }
        }
        
        EPD_10in85g_SendCommand_1(0x10);
        for (UDOUBLE i = 0; i < Height; i++) {
            for (UDOUBLE j = 0; j < Width; j++) {
                EPD_10in85g_SendData_1(0x55);
            }
        }
    }
    else
    {
        EPD_10in85g_SendCommand_0(0x10);
        for (UDOUBLE i = 0; i < Height; i++) {
            for (UDOUBLE j = 0; j < Width; j++) {
                if((i<Yend) && (i>=ystart) && (j>=xstart)) {
                    EPD_10in85g_SendData_0(Image[(j-xstart) + (image_width/4*(i-ystart))]);
                }
                else
                    EPD_10in85g_SendData_0(0x55);
            }
        }

        EPD_10in85g_SendCommand_1(0x10);
        for (UDOUBLE i = 0; i < Height; i++) {
            for (UDOUBLE j = 0; j < Width; j++) {
                if((i<Yend) && (i>=ystart) && (j<Xend-170)) {
                    EPD_10in85g_SendData_1(Image[(j+170-xstart) + (image_width/4*(i-ystart))]);
                }
                else
                    EPD_10in85g_SendData_1(0x55);
            }
        }
    }

    EPD_10in85g_TurnOnDisplay();
}

/******************************************************************************
function :	Enter sleep mode
parameter:
******************************************************************************/
void EPD_10in85g_Sleep(void)
{
    EPD_10in85g_SendCommand_ALL(0x02); 
    EPD_10in85g_SendData_ALL(0x00); 
	DEV_Delay_ms(100);  
	EPD_10in85g_ReadBusy();       

	EPD_10in85g_SendCommand_ALL(0X07); //enter deep sleep
	EPD_10in85g_SendData_ALL(0xA5); 
	DEV_Delay_ms(100);
}
